src/foam/nanos/rope/ROPEAuthorizer.js

```diff
@@ -0,0 +1,224 @@
+        {
+          name: 'obj',
+          javaType: 'FObject'
+        }
+      ],
+      javaCode: `
+      if ( ((User) x.get("user")).getId() == User.SYSTEM_USER_ID ) return;
+  
+      List<String> propertiesUpdated = getPropertiesUpdated(null, obj);
+      for ( String property : propertiesUpdated ) {
+        if ( ! authorizeByRope(x, obj, getTargetDAOKey(), "create", property ) ) {
+          throw new AuthorizationException("You don't have permission to create this object");
+        }
+      }
+      `
+    },
+    {
+      name: 'authorizeOnRead',
+      type: 'Void',
+      javaThrows: [
+        'AuthorizationException'
+      ],
+      args: [
+        {
+          name: 'x',
+          javaType: 'X'
+        },
+        {
+          name: 'obj',
+          javaType: 'FObject'
+        }
+      ],
+      javaCode: `
+      if ( ((User) x.get("user")).getId() == User.SYSTEM_USER_ID ) return;
+      if ( ! authorizeByRope(x, obj, getTargetDAOKey(), "read", "") ) throw new AuthorizationException("You don't have permission to read this object");
+      `
+    },
+    {
+      name: 'authorizeOnUpdate',
+      type: 'Void',
+      javaThrows: [
+        'AuthorizationException'
+      ],
+      args: [
+        {
+          name: 'x',
+          type: 'foam.core.X'
+        },
+        {
+          name: 'oldObj',
+          type: 'foam.core.FObject'
+        },
+        {
+          name: 'newObj',
+          type: 'foam.core.FObject'
+        },
+      ],
+      javaCode: `
+      if ( ((User) x.get("user")).getId() == User.SYSTEM_USER_ID ) return;
+  
+      List<String> propertiesUpdated = getPropertiesUpdated(oldObj, newObj);
+      for ( String property : propertiesUpdated ) {
+        if ( ! authorizeByRope(x, newObj, getTargetDAOKey(), "update", property ) ) {
+          throw new AuthorizationException("You don't have permission to update this object");
+        }
+      }
+      `
+    },
+    {
+      name: 'authorizeOnDelete',
+      type: 'Void',
+      javaThrows: [
+        'AuthorizationException'
+      ],
+      args: [
+        {
+          name: 'x',
+          javaType: 'X'
+        },
+        {
+          name: 'obj',
+          javaType: 'FObject'
+        }
+      ],
+      javaCode: `
+      if ( ((User) x.get("user")).getId() == User.SYSTEM_USER_ID ) return;
+      if ( ! authorizeByRope(x, obj, getTargetDAOKey(), "delete", "") ) throw new AuthorizationException("You don't have permission to delete this object");
+      `
+    },
+    {
+      name: 'checkGlobalRead',
+      javaType: 'boolean',
+      args: [
+        {
+          name: 'x',
+          javaType: 'X'
+        }
+      ],
+      javaCode: `
+      return false;
+      `
+    },
+    {
+      name: 'checkGlobalRemove',
+      javaType: 'boolean',
+      args: [
+        {
+          name: 'x',
+          javaType: 'X'
+        }
+      ],
+      javaCode: `
+      return false;
+      `
+    },
+    {
+      name: 'getPermissionPrefix',
```

Comment:
This isn't a method on the `Authorizer` interface. Do we need it?

Reply:
this got added into the authorizer interface a while ago, but its not really being used for anything. can i remove from interface?

Reply:
I see. Yeah, I don't think it makes sense there.